File: internal/db/knowledge.go
package db

import (
	"database/sql"
	"encoding/json"
	"strings"
	"time"
	"unicode"
)

const (
	KnowledgeSourceManual = "manual"
	KnowledgeSourceUpload = "upload"

	defaultKnowledgeLimit = 5
	maxKnowledgeLimit     = 10
	maxChunkRunes         = 900
	minChunkRunes         = 120
)

type KnowledgeBase struct {
	ID          int64     `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

type KnowledgeDocument struct {
	ID              int64     `json:"id"`
	KnowledgeBaseID int64     `json:"knowledge_base_id"`
	Title           string    `json:"title"`
	Content         string    `json:"content"`
	Tags            []string  `json:"tags"`
	SourceType      string    `json:"source_type"`
	SourceName      string    `json:"source_name"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

type KnowledgeDocumentFilter struct {
	KnowledgeBaseID int64  `json:"knowledge_base_id"`
	Query           string `json:"query"`
}

type KnowledgeSearchFilter struct {
	Query           string `json:"query"`
	KnowledgeBaseID int64  `json:"knowledge_base_id"`
	Limit           int    `json:"limit"`
}

type KnowledgeSearchResult struct {
	KnowledgeBaseID   int64   `json:"knowledge_base_id"`
	KnowledgeBaseName string  `json:"knowledge_base_name"`
	DocumentID        int64   `json:"document_id"`
	DocumentTitle     string  `json:"document_title"`
	ChunkID           int64   `json:"chunk_id"`
	Snippet           string  `json:"snippet"`
	Score             float64 `json:"score"`
}

// CreateKnowledgeBase inserts a new knowledge base.
func (db *Database) CreateKnowledgeBase(base *KnowledgeBase) error {
	now := time.Now()
	res, err := db.conn.Exec(
		`INSERT INTO knowledge_bases (name, description, created_at, updated_at) VALUES (?, ?, ?, ?)`,
		base.Name, base.Description, now, now,
	)
	if err != nil {
		return err
	}
	base.ID, _ = res.LastInsertId()
	base.CreatedAt = now
	base.UpdatedAt = now
	return nil
}

// ListKnowledgeBases lists all knowledge bases.
func (db *Database) ListKnowledgeBases() ([]KnowledgeBase, error) {
	rows, err := db.conn.Query(`SELECT id, name, description, created_at, updated_at FROM knowledge_bases ORDER BY updated_at DESC, id DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var bases []KnowledgeBase
	for rows.Next() {
		var base KnowledgeBase
		if err := rows.Scan(&base.ID, &base.Name, &base.Description, &base.CreatedAt, &base.UpdatedAt); err != nil {
			return nil, err
		}
		bases = append(bases, base)
	}
	return bases, rows.Err()
}

// GetKnowledgeBase retrieves a single knowledge base by ID.
func (db *Database) GetKnowledgeBase(id int64) (*KnowledgeBase, error) {
	var base KnowledgeBase
	err := db.conn.QueryRow(
		`SELECT id, name, description, created_at, updated_at FROM knowledge_bases WHERE id = ?`, id,
	).Scan(&base.ID, &base.Name, &base.Description, &base.CreatedAt, &base.UpdatedAt)
	if err != nil {
		return nil, err
	}
	return &base, nil
}

// UpdateKnowledgeBase updates an existing knowledge base.
func (db *Database) UpdateKnowledgeBase(base *KnowledgeBase) error {
	now := time.Now()
	res, err := db.conn.Exec(
		`UPDATE knowledge_bases SET name = ?, description = ?, updated_at = ? WHERE id = ?`,
		base.Name, base.Description, now, base.ID,
	)
	if err != nil {
		return err
	}
	if err := errNoRowsWhenUnchanged(res); err != nil {
		return err
	}
	base.UpdatedAt = now
	return nil
}

// DeleteKnowledgeBase removes a knowledge base and cascades its documents and chunks.
func (db *Database) DeleteKnowledgeBase(id int64) error {
	tx, err := db.conn.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	if _, err := tx.Exec(`DELETE FROM knowledge_chunks_fts WHERE knowledge_base_id = ?`, id); err != nil {
		return err
	}
	res, err := tx.Exec(`DELETE FROM knowledge_bases WHERE id = ?`, id)
	if err != nil {
		return err
	}
	if err := errNoRowsWhenUnchanged(res); err != nil {
		return err
	}
	return tx.Commit()
}

// CreateKnowledgeDocument inserts a document and refreshes searchable chunks.
func (db *Database) CreateKnowledgeDocument(doc *KnowledgeDocument) error {
	tx, err := db.conn.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	tagsJSON, err := marshalKnowledgeTags(doc.Tags)
	if err != nil {
		return err
	}
	doc.SourceType = defaultKnowledgeSourceType(doc.SourceType)
	now := time.Now()
	res, err := tx.Exec(
		`INSERT INTO knowledge_documents (knowledge_base_id, title, content, tags, source_type, source_name, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
		doc.KnowledgeBaseID, doc.Title, doc.Content, tagsJSON, doc.SourceType, doc.SourceName, now, now,
	)
	if err != nil {
		return err
	}
	doc.ID, _ = res.LastInsertId()
	doc.CreatedAt = now
	doc.UpdatedAt = now
	if err := refreshKnowledgeChunks(tx, doc.ID, doc.KnowledgeBaseID, doc.Content); err != nil {
		return err
	}
	return tx.Commit()
}

// ListKnowledgeDocuments lists documents with optional base and text filters.
func (db *Database) ListKnowledgeDocuments(filter KnowledgeDocumentFilter) ([]KnowledgeDocument, error) {
	query := `SELECT id, knowledge_base_id, title, content, tags, source_type, source_name, created_at, updated_at FROM knowledge_documents`
	var args []interface{}
	var where []string
	if filter.KnowledgeBaseID > 0 {
		where = append(where, "knowledge_base_id = ?")
		args = append(args, filter.KnowledgeBaseID)
	}
	if q := strings.TrimSpace(filter.Query); q != "" {
		where = append(where, "(title LIKE ? OR content LIKE ? OR tags LIKE ?)")
		like := "%" + q + "%"
		args = append(args, like, like, like)
	}
	if len(where) > 0 {
		query += " WHERE " + strings.Join(where, " AND ")
	}
	query += ` ORDER BY updated_at DESC, id DESC`

	rows, err := db.conn.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var docs []KnowledgeDocument
	for rows.Next() {
		doc, err := scanKnowledgeDocument(rows)
		if err != nil {
			return nil, err
		}
		docs = append(docs, *doc)
	}
	return docs, rows.Err()
}

// GetKnowledgeDocument retrieves a single document by ID.
func (db *Database) GetKnowledgeDocument(id int64) (*KnowledgeDocument, error) {
	row := db.conn.QueryRow(
		`SELECT id, knowledge_base_id, title, content, tags, source_type, source_name, created_at, updated_at FROM knowledge_documents WHERE id = ?`, id,
	)
	return scanKnowledgeDocument(row)
}

// UpdateKnowledgeDocument updates a document and refreshes searchable chunks.
func (db *Database) UpdateKnowledgeDocument(doc *KnowledgeDocument) error {
	tx, err := db.conn.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	tagsJSON, err := marshalKnowledgeTags(doc.Tags)
	if err != nil {
		return err
	}
	doc.SourceType = defaultKnowledgeSourceType(doc.SourceType)
	now := time.Now()
	res, err := tx.Exec(
		`UPDATE knowledge_documents SET knowledge_base_id = ?, title = ?, content = ?, tags = ?, source_type = ?, source_name = ?, updated_at = ? WHERE id = ?`,
		doc.KnowledgeBaseID, doc.Title, doc.Content, tagsJSON, doc.SourceType, doc.SourceName, now, doc.ID,
	)
	if err != nil {
		return err
	}
	if err := errNoRowsWhenUnchanged(res); err != nil {
		return err
	}
	if err := refreshKnowledgeChunks(tx, doc.ID, doc.KnowledgeBaseID, doc.Content); err != nil {
		return err
	}
	doc.UpdatedAt = now
	return tx.Commit()
}

// DeleteKnowledgeDocument removes a document and its FTS rows.
func (db *Database) DeleteKnowledgeDocument(id int64) error {
	tx, err := db.conn.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	if _, err := tx.Exec(`DELETE FROM knowledge_chunks_fts WHERE document_id = ?`, id); err != nil {
		return err
	}
	res, err := tx.Exec(`DELETE FROM knowledge_documents WHERE id = ?`, id)
	if err != nil {
		return err
	}
	if err := errNoRowsWhenUnchanged(res); err != nil {
		return err
	}
	return tx.Commit()
}

// SearchKnowledge searches knowledge chunks through SQLite FTS.
func (db *Database) SearchKnowledge(filter KnowledgeSearchFilter) ([]KnowledgeSearchResult, error) {
	queryText := strings.TrimSpace(filter.Query)
	if queryText == "" {
		return []KnowledgeSearchResult{}, nil
	}
	limit := filter.Limit
	if limit <= 0 {
		limit = defaultKnowledgeLimit
	}
	if limit > maxKnowledgeLimit {
		limit = maxKnowledgeLimit
	}

	query := `SELECT f.knowledge_base_id, b.name, f.document_id, d.title, f.chunk_id,
			snippet(knowledge_chunks_fts, 3, '', '', '...', 20), bm25(knowledge_chunks_fts) AS score
		FROM knowledge_chunks_fts f
		JOIN knowledge_documents d ON d.id = f.document_id
		JOIN knowledge_bases b ON b.id = f.knowledge_base_id
		WHERE knowledge_chunks_fts MATCH ?`
	args := []interface{}{buildKnowledgeFTSQuery(queryText)}
	if filter.KnowledgeBaseID > 0 {
		query += ` AND f.knowledge_base_id = ?`
		args = append(args, filter.KnowledgeBaseID)
	}
	query += ` ORDER BY score ASC LIMIT ?`
	args = append(args, limit)

	rows, err := db.conn.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []KnowledgeSearchResult
	for rows.Next() {
		var result KnowledgeSearchResult
		if err := rows.Scan(&result.KnowledgeBaseID, &result.KnowledgeBaseName, &result.DocumentID, &result.DocumentTitle, &result.ChunkID, &result.Snippet, &result.Score); err != nil {
			return nil, err
		}
		results = append(results, result)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	if len(results) > 0 {
		return results, nil
	}
	return db.searchKnowledgeByText(filter, queryText, limit)
}

type knowledgeDocumentScanner interface {
	Scan(dest ...interface{}) error
}

func scanKnowledgeDocument(scanner knowledgeDocumentScanner) (*KnowledgeDocument, error) {
	var doc KnowledgeDocument
	var tagsJSON string
	if err := scanner.Scan(&doc.ID, &doc.KnowledgeBaseID, &doc.Title, &doc.Content, &tagsJSON, &doc.SourceType, &doc.SourceName, &doc.CreatedAt, &doc.UpdatedAt); err != nil {
		return nil, err
	}
	if tagsJSON != "" {
		if err := json.Unmarshal([]byte(tagsJSON), &doc.Tags); err != nil {
			return nil, err
		}
	}
	return &doc, nil
}

func marshalKnowledgeTags(tags []string) (string, error) {
	if tags == nil {
		tags = []string{}
	}
	data, err := json.Marshal(tags)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func defaultKnowledgeSourceType(sourceType string) string {
	if strings.TrimSpace(sourceType) == "" {
		return KnowledgeSourceManual
	}
	return sourceType
}

func refreshKnowledgeChunks(tx *sql.Tx, documentID, knowledgeBaseID int64, content string) error {
	if _, err := tx.Exec(`DELETE FROM knowledge_chunks_fts WHERE document_id = ?`, documentID); err != nil {
		return err
	}
	if _, err := tx.Exec(`DELETE FROM knowledge_chunks WHERE document_id = ?`, documentID); err != nil {
		return err
	}
	for i, chunk := range chunkKnowledgeContent(content) {
		res, err := tx.Exec(
			`INSERT INTO knowledge_chunks (document_id, knowledge_base_id, chunk_index, content) VALUES (?, ?, ?, ?)`,
			documentID, knowledgeBaseID, i, chunk,
		)
		if err != nil {
			return err
		}
		chunkID, _ := res.LastInsertId()
		if _, err := tx.Exec(
			`INSERT INTO knowledge_chunks_fts (chunk_id, document_id, knowledge_base_id, content) VALUES (?, ?, ?, ?)`,
			chunkID, documentID, knowledgeBaseID, chunk,
		); err != nil {
			return err
		}
	}
	return nil
}

func chunkKnowledgeContent(content string) []string {
	paragraphs := splitKnowledgeParagraphs(strings.TrimSpace(content))
	if len(paragraphs) == 0 {
		return nil
	}
	paragraphs = mergeShortKnowledgeParagraphs(paragraphs)

	var chunks []string
	for _, paragraph := range paragraphs {
		chunks = append(chunks, splitLongKnowledgeChunk(paragraph, maxChunkRunes)...)
	}
	return chunks
}

func splitKnowledgeParagraphs(content string) []string {
	var paragraphs []string
	var current []string
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			if len(current) > 0 {
				paragraphs = append(paragraphs, strings.Join(current, "\n"))
				current = nil
			}
			continue
		}
		current = append(current, line)
	}
	if len(current) > 0 {
		paragraphs = append(paragraphs, strings.Join(current, "\n"))
	}
	return paragraphs
}

func mergeShortKnowledgeParagraphs(paragraphs []string) []string {
	var merged []string
	for _, paragraph := range paragraphs {
		paragraph = strings.TrimSpace(paragraph)
		if paragraph == "" {
			continue
		}
		if len(merged) > 0 && runeLen(merged[len(merged)-1]) < minChunkRunes && runeLen(merged[len(merged)-1])+runeLen(paragraph)+2 <= maxChunkRunes {
			merged[len(merged)-1] += "\n\n" + paragraph
			continue
		}
		merged = append(merged, paragraph)
	}
	return merged
}

func splitLongKnowledgeChunk(content string, maxRunes int) []string {
	runes := []rune(strings.TrimSpace(content))
	if len(runes) <= maxRunes {
		return []string{string(runes)}
	}

	var chunks []string
	for len(runes) > maxRunes {
		cut := maxRunes
		for cut > maxRunes/2 && !unicode.IsSpace(runes[cut-1]) {
			cut--
		}
		if cut <= maxRunes/2 {
			cut = maxRunes
		}
		chunk := strings.TrimSpace(string(runes[:cut]))
		if chunk != "" {
			chunks = append(chunks, chunk)
		}
		runes = []rune(strings.TrimSpace(string(runes[cut:])))
	}
	if len(runes) > 0 {
		chunks = append(chunks, string(runes))
	}
	return chunks
}

func runeLen(s string) int {
	return len([]rune(s))
}

func buildKnowledgeFTSQuery(query string) string {
	var terms []string
	for _, field := range strings.Fields(query) {
		field = strings.ReplaceAll(field, `"`, `""`)
		if field != "" {
			terms = append(terms, `"`+field+`"`)
		}
	}
	if len(terms) == 0 {
		return `""`
	}
	return strings.Join(terms, " ")
}

func (db *Database) searchKnowledgeByText(filter KnowledgeSearchFilter, queryText string, limit int) ([]KnowledgeSearchResult, error) {
	patterns := buildKnowledgeLikePatterns(queryText)
	if len(patterns) == 0 {
		return []KnowledgeSearchResult{}, nil
	}

	var clauses []string
	var args []interface{}
	for _, pattern := range patterns {
		clauses = append(clauses, "(c.content LIKE ? OR d.title LIKE ?)")
		like := "%" + pattern + "%"
		args = append(args, like, like)
	}

	query := `SELECT c.knowledge_base_id, b.name, c.document_id, d.title, c.id, c.content, 0.0 AS score
		FROM knowledge_chunks c
		JOIN knowledge_documents d ON d.id = c.document_id
		JOIN knowledge_bases b ON b.id = c.knowledge_base_id
		WHERE (` + strings.Join(clauses, " OR ") + `)`
	if filter.KnowledgeBaseID > 0 {
		query += ` AND c.knowledge_base_id = ?`
		args = append(args, filter.KnowledgeBaseID)
	}
	query += ` ORDER BY d.updated_at DESC, c.chunk_index ASC LIMIT ?`
	args = append(args, limit)

	rows, err := db.conn.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []KnowledgeSearchResult
	for rows.Next() {
		var result KnowledgeSearchResult
		if err := rows.Scan(&result.KnowledgeBaseID, &result.KnowledgeBaseName, &result.DocumentID, &result.DocumentTitle, &result.ChunkID, &result.Snippet, &result.Score); err != nil {
			return nil, err
		}
		results = append(results, result)
	}
	return results, rows.Err()
}

func buildKnowledgeLikePatterns(query string) []string {
	seen := map[string]bool{}
	var patterns []string
	add := func(pattern string) {
		pattern = strings.TrimSpace(pattern)
		if pattern == "" || seen[pattern] {
			return
		}
		seen[pattern] = true
		patterns = append(patterns, pattern)
	}

	add(query)
	for _, field := range strings.Fields(query) {
		add(field)
	}
	for _, segment := range cjkSegments(query) {
		runes := []rune(segment)
		for width := minInt(8, len(runes)); width >= 2; width-- {
			for start := 0; start+width <= len(runes); start++ {
				add(string(runes[start : start+width]))
				if len(patterns) >= 24 {
					return patterns
				}
			}
		}
	}
	return patterns
}

func cjkSegments(s string) []string {
	var segments []string
	var current []rune
	flush := func() {
		if len(current) > 0 {
			segments = append(segments, string(current))
			current = nil
		}
	}
	for _, r := range s {
		if unicode.Is(unicode.Han, r) {
			current = append(current, r)
			continue
		}
		flush()
	}
	flush()
	return segments
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: src/offerpilot/repositories/knowledge.py
from __future__ import annotations

from dataclasses import dataclass
import re
from typing import Any, Optional

from sqlalchemy import delete, or_, select, text
from sqlalchemy.orm import Session, sessionmaker

from offerpilot.models import KnowledgeChunk, KnowledgeDocument


@dataclass
class KnowledgeDocumentCreate:
    title: str
    content: str = ""
    tags: list[str] | None = None
    source_type: str = "manual"
    source_name: str = ""


class KnowledgeRepository:
    def __init__(self, session_factory: sessionmaker[Session]):
        self._session_factory = session_factory

    def create_document(self, data: KnowledgeDocumentCreate) -> KnowledgeDocument:
        doc = KnowledgeDocument(
            title=data.title,
            content=data.content,
            source_type=data.source_type or "manual",
            source_name=data.source_name,
        )
        doc.tags = data.tags or []
        with self._session_factory() as session:
            session.add(doc)
            session.flush()
            _refresh_chunks(session, doc.id, doc.content)
            session.commit()
            session.refresh(doc)
            return doc

    def list_documents(self, query: str = "") -> list[KnowledgeDocument]:
        statement = select(KnowledgeDocument)
        if query.strip():
            like = f"%{query.strip()}%"
            statement = statement.where(
                or_(
                    KnowledgeDocument.title.like(like),
                    KnowledgeDocument.content.like(like),
                    KnowledgeDocument._tags.like(like),
                )
            )
        statement = statement.order_by(KnowledgeDocument.updated_at.desc(), KnowledgeDocument.id.desc())
        with self._session_factory() as session:
            return list(session.scalars(statement))

    def get_document(self, document_id: int) -> Optional[KnowledgeDocument]:
        with self._session_factory() as session:
            return session.get(KnowledgeDocument, document_id)

    def update_document(
        self,
        document_id: int,
        data: KnowledgeDocumentCreate,
    ) -> Optional[KnowledgeDocument]:
        with self._session_factory() as session:
            doc = session.get(KnowledgeDocument, document_id)
            if doc is None:
                return None
            doc.title = data.title
            doc.content = data.content
            doc.tags = data.tags or []
            doc.source_type = data.source_type or doc.source_type
            doc.source_name = data.source_name
            _refresh_chunks(session, doc.id, doc.content)
            session.commit()
            session.refresh(doc)
            return doc

    def delete_document(self, document_id: int) -> bool:
        with self._session_factory() as session:
            doc = session.get(KnowledgeDocument, document_id)
            if doc is None:
                return False
            _delete_fts_for_document(session, document_id)
            session.delete(doc)
            session.commit()
            return True

    def search(self, query: str, limit: int = 5) -> list[dict[str, Any]]:
        limit = min(max(limit or 5, 1), 10)
        terms = _search_patterns(query)
        if not terms:
            return []
        with self._session_factory() as session:
            fts_rows = _search_fts(session, query, limit * 3)
            lexical_rows = _search_lexical(session, terms, limit * 3)
        return _merge_ranked_results([fts_rows, lexical_rows], limit)


def _search_lexical(
    session: Session,
    terms: list[str],
    limit: int,
) -> list[dict[str, Any]]:
    statement = (
        select(KnowledgeChunk, KnowledgeDocument)
        .join(KnowledgeDocument, KnowledgeDocument.id == KnowledgeChunk.document_id)
    )
    statement = statement.where(
        or_(
            *[
                or_(
                    KnowledgeChunk.content.like(f"%{term}%"),
                    KnowledgeDocument.title.like(f"%{term}%"),
                )
                for term in terms
            ]
        )
    )
    statement = statement.order_by(KnowledgeDocument.updated_at.desc(), KnowledgeChunk.chunk_index.asc()).limit(limit)
    rows = session.execute(statement).all()
    return [_search_payload(chunk, doc) for chunk, doc in rows]


def _search_fts(
    session: Session,
    query: str,
    limit: int,
) -> list[dict[str, Any]]:
    fts_query = _fts_query(query)
    if not fts_query:
        return []
    where = "knowledge_chunks_fts MATCH :query"
    params: dict[str, Any] = {"query": fts_query, "limit": limit}
    try:
        rows = session.execute(
            text(
                f"""
                SELECT
                    c.id AS chunk_id,
                    c.chunk_index AS chunk_index,
                    c.content AS snippet,
                    d.id AS document_id,
                    d.title AS document_title,
                    d.source_name AS source_name,
                    bm25(knowledge_chunks_fts) AS rank
                FROM knowledge_chunks_fts
                JOIN knowledge_chunks c ON c.id = knowledge_chunks_fts.chunk_id
                JOIN knowledge_documents d ON d.id = c.document_id
                WHERE {where}
                ORDER BY rank ASC
                LIMIT :limit
                """
            ),
            params,
        ).mappings()
    except Exception:
        return []
    return [
        {
            "document_id": int(row["document_id"]),
            "document_title": str(row["document_title"]),
            "source_name": str(row["source_name"] or ""),
            "chunk_id": int(row["chunk_id"]),
            "chunk_index": int(row["chunk_index"]),
            "snippet": str(row["snippet"]),
            "score": 0.0,
        }
        for row in rows
    ]


def _merge_ranked_results(result_sets: list[list[dict[str, Any]]], limit: int) -> list[dict[str, Any]]:
    merged: dict[int, dict[str, Any]] = {}
    for result_set in result_sets:
        for rank, row in enumerate(result_set, start=1):
            chunk_id = int(row["chunk_id"])
            if chunk_id not in merged:
                merged[chunk_id] = dict(row)
            merged[chunk_id]["score"] = float(merged[chunk_id]["score"]) + (1.0 / (60 + rank))
    ranked = sorted(
        merged.values(),
        key=lambda row: (-float(row["score"]), int(row["document_id"]), int(row["chunk_index"])),
    )
    for row in ranked:
        row["score"] = round(float(row["score"]), 6)
    return ranked[:limit]


def _search_payload(
    chunk: KnowledgeChunk,
    doc: KnowledgeDocument,
) -> dict[str, Any]:
    return {
        "document_id": doc.id,
        "document_title": doc.title,
        "source_name": doc.source_name,
        "chunk_id": chunk.id,
        "chunk_index": chunk.chunk_index,
        "snippet": chunk.content,
        "score": 0.0,
    }


def _refresh_chunks(session: Session, document_id: int, content: str) -> None:
    session.execute(delete(KnowledgeChunk).where(KnowledgeChunk.document_id == document_id))
    _delete_fts_for_document(session, document_id)
    chunks: list[KnowledgeChunk] = []
    for index, chunk in enumerate(_chunk_content(content)):
        chunks.append(
            KnowledgeChunk(
                document_id=document_id,
                chunk_index=index,
                content=chunk,
            )
        )
    session.add_all(chunks)
    session.flush()
    _insert_fts_chunks(session, chunks)


def _insert_fts_chunks(session: Session, chunks: list[KnowledgeChunk]) -> None:
    if not chunks:
        return
    try:
        for chunk in chunks:
            session.execute(
                text(
                    """
                    INSERT INTO knowledge_chunks_fts
                        (chunk_id, document_id, content)
                    VALUES (:chunk_id, :document_id, :content)
                    """
                ),
                {
                    "chunk_id": chunk.id,
                    "document_id": chunk.document_id,
                    "content": chunk.content,
                },
            )
    except Exception:
        return


def _delete_fts_for_document(session: Session, document_id: int) -> None:
    try:
        session.execute(
            text("DELETE FROM knowledge_chunks_fts WHERE document_id = :document_id"),
            {"document_id": document_id},
        )
    except Exception:
        return

def _chunk_content(content: str) -> list[str]:
    stripped = content.strip()
    if not stripped:
        return []
    chunks = [part.strip() for part in stripped.split("\n\n") if part.strip()]
    return chunks or [stripped]


def _search_patterns(query: str) -> list[str]:
    seen: set[str] = set()
    patterns: list[str] = []
    for term in [query.strip(), *query.split()]:
        term = term.strip()
        if term and term not in seen:
            seen.add(term)
            patterns.append(term)
    return patterns


def _fts_query(query: str) -> str:
    tokens = re.findall(r"[\w\u4e00-\u9fff]+", query.lower())
    deduped = list(dict.fromkeys(token for token in tokens if token))
    return " OR ".join(f'"{token}"' for token in deduped)
